### PR TITLE
64 bit: Initial support for Linux x86-64 native.

### PR DESCRIPTION
**Pull request by fzi-haxel**
### Contribution description

Initial support for 64 bit in the kernel, pkgs, examples, tests and the native board.

#### Why add 64 bit support?

As mentioned in the FAQ on www.riot-os.org, there are some benefits
* Improve RIOT’s code base by fixing 64 bit related errors
* Simplify toolchain setup for native 
    * Users can compile native without having to install additional 32 bit libraries

The FAQ also mentions some cons
* Effort / gain (as it’s unlikely we will support any other 64 bit platform, fixing potential 64 bit errors in RIOT is less beneficial)
    * The effort for basic 64 bit support was not very large. The kernel itself is quite well written, as it already supports different architectures. The biggest difference in 64 bit is that `sizeof(void*) == sizeof(size_t) != sizeof(int)`, which leads to quite a few compiler warnings in printf formatting and casting between integer values and pointers. This is especially true for the tests. Other than that, and 64 bit support for the native board, the rest of the changes are fairly minor. I also tried to change as little as possible to avoid breaking compatibility with the other boards. See the commit messages for more details.
    * I agree that RIOT should focus on smaller architectures. However, I don't think this has to prevent RIOT from also supporting other 64 bit boards, such as riscv64-based ones.
* Decreased 32 bit testing (unless we make 32 the default in which case the toolchain setup argument is diminished)
    * I definitely agree that 32 bit native should remain a default platform for testing, especially in the CI. But I don't see much of a downside if a user tests their application with 64 bit native (especially if it's not for a PR to the RIOT repo).

#### Native for Linux x86-64

For the initial version, instead of creating a separate `native64` board, I decided to add the variable `NATIVE_64BIT` to the native board Makefiles. Setting the environment to `NATIVE_64BIT=y` will enable 64-bit mode for native, which is disabled by default.

While I don't necessarily think this is the best long-term solution, it does ensure that all tests run with the same configuration as the 32 bit version. A separate board would require a major refactoring of many tests, which often have different configurations in the Makefile, C code, and Python tests if the board is native, and I didn't want to increase the size of an already large pull request.

Almost all of the same features as the 32 bit version are supported by the 64 bit version, but the following are still missing 
* Architectures other than x86-64 or operating systems other than Linux
    * No FreeBSD support
    * No Aarch64 support
* Rust support for x86-64

#### Some closing thoughts

This PR is a suggestion for initial 64 bit support, feel free to suggest any changes.
I thought about making several smaller PRs, but that would have made it difficult to test the 64 bit part.
So I decided to go for the *big bang* solution. If you have a better suggestion, I'm open to restructuring the PR.

**DISCLAMER:** While I did a more thorough review of the kernel, I did not do a thorough review of the entire code base. I mainly fixed compilation or testing errors that came up during testing. I am sure I have missed at least some DEBUG/printf formatting issues. There is also some optimization potential for 64 bit, but I focused on minimal changes.

### Testing procedure

Just set the `NATIVE_64BIT=y` environment variable and compile something for native.

For example
`NATIVE_64BIT=y ./dist/tools/compile_and_test_for_board/compile_and_test_for_board.py . native`

**NOTE:**
* Setting `NATIVE_64BIT=y` will skip the unsupported Rust examples and tests. Other than that, nothing else should be skipped.
* 64 bit native may require additional dependencies. The CI is missing `libsdl2-dev` for 64 bit.

### Issues/PRs references

See also #6603

### CODE REVIEW

File: pkg/flashdb/patches/0002-64bit-fix.patch [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+diff --git a/src/fdb_kvdb.c b/src/fdb_kvdb.c
+index 393c463..7905fda 100644
+--- a/src/fdb_kvdb.c
++++ b/src/fdb_kvdb.c
+@@ -341,7 +341,7 @@ static fdb_err_t read_kv(fdb_kvdb_t db, fdb_kv_t kv)
+     kv->status = (fdb_kv_status_t) _fdb_get_status(kv_hdr.status_table, FDB_KV_STATUS_NUM);
+     kv->len = kv_hdr.len;
+ 
+-    if (kv->len == ~0UL || kv->len > db_max_size(db) || kv->len < KV_HDR_DATA_SIZE) {
++    if (kv->len == UINT32_MAX || kv->len > db_max_size(db) || kv->len < KV_HDR_DATA_SIZE) {
```
Comment (benpicco) [inline pkg/flashdb/patches/0002-64bit-fix.patch:19]: [upstream](https://github.com/armink/FlashDB) is pretty responsive when it comes to patches, you can just open a PR there and we can update the commit hash.

File: pkg/ccn-lite/patches/0007-print-fix-for-64-bit.patch [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+diff --git a/src/ccnl-riot/src/ccn-lite-riot.c b/src/ccnl-riot/src/ccn-lite-riot.c
+index fde0a44..0ddb488 100644
+--- a/src/ccnl-riot/src/ccn-lite-riot.c
++++ b/src/ccnl-riot/src/ccn-lite-riot.c
+@@ -278,7 +278,7 @@ int
+ ccnl_app_RX(struct ccnl_relay_s *ccnl, struct ccnl_content_s *c)
+ {
+     (void) ccnl;
+-    DEBUGMSG(DEBUG, "Received something of size %u for the application\n", c->pkt->contlen);
++    DEBUGMSG(DEBUG, "Received something of size %u for the application\n", (unsigned)c->pkt->contlen);
```
Comment (benpicco) [inline pkg/ccn-lite/patches/0007-print-fix-for-64-bit.patch:19]: You can also move that upstream

**Review by benpicco — COMMENTED** [review]
If you move the preparatory patches (64 bit safety compile fixes) to a separate PR we can merge those quickly.

For changes to third party packages, please try to also make a PR upstream so we can drop the downstream patch next time we update the pkt version.

Then you'd only have changes to `native` left in this PR, which makes review easier (and reduces the chance of merge conflicts).

**Comment by OlegHahm** [comment]
Is this based on #13009 or a complete rework?

**Comment by fzi-haxel** [comment]
> Is this based on #13009 or a complete rework?

No, it's technically based on this [fork](https://github.com/fzi-forschungszentrum-informatik/RIOT64) that we did a while ago.

**Comment by MrKevinWeiss** [comment]
probably this won't make it into the release.

**Comment by fzi-haxel** [comment]
I changed the PR to WiP and updated the commits so people can still test 64bit native.

Updates include:
- Updated to current discussion in #20154 
- Rebased to master
- Some cleanup in the individual commits and commit messages

**Comment by fzi-haxel** [comment]
Closing this PR. The PRs above have been merged.

**Pull request closed without merging**